src/components/inputs/TokenInput/TokenInput.vue

```diff
@@ -0,0 +1,219 @@
+import { HtmlInputEvent } from '@/types';
+import { ref, computed, watchEffect } from 'vue';
+import useTokens from '@/composables/useTokens';
+import TokenSelectInput from '@/components/inputs/TokenSelectInput/TokenSelectInput.vue';
+import useNumbers from '@/composables/useNumbers';
+import { bnum } from '@/lib/utils';
+import useUserSettings from '@/composables/useUserSettings';
+import { isPositive, isLessThanOrEqualTo } from '@/lib/utils/validations';
+import { useI18n } from 'vue-i18n';
+import useWeb3 from '@/services/web3/useWeb3';
+
+/**
+ * TYPES
+ */
+type InputValue = string | number;
+
+type Props = {
+  amount: InputValue;
+  address?: string;
+  weight?: number;
+  noRules?: boolean;
+  noMax?: boolean;
+  priceImpact?: number;
+  label?: string;
+  fixedToken?: boolean;
+  customBalance?: string;
+  disableMax?: boolean;
+};
+
+/**
+ * PROPS & EMITS
+ */
+const props = withDefaults(defineProps<Props>(), {
+  amount: '',
+  address: '',
+  weight: 0,
+  noRules: false,
+  noMax: false,
+  fixedToken: false,
+  disableMax: false
+});
+
+const emit = defineEmits<{
+  (e: 'blur', value: string): void;
+  (e: 'input', value: string): void;
+  (e: 'update:amount', value: string): void;
+  (e: 'update:address', value: string): void;
+  (e: 'update:isValid', value: boolean): void;
+  (e: 'keydown', value: HtmlInputEvent);
+}>();
+
+/**
+ * STATE
+ */
+const _amount = ref<InputValue>('');
+const _address = ref<string>('');
+const ETH_BUFFER = 0.1;
+
+/**
+ * COMPOSABLEs
+ */
+const { getToken, balanceFor, nativeAsset } = useTokens();
+const { fNum, toFiat } = useNumbers();
+const { currency } = useUserSettings();
+const { t } = useI18n();
+const { isWalletReady } = useWeb3();
+
+/**
+ * COMPUTED
+ */
+const hasToken = computed(() => !!_address.value);
+const hasAmount = computed(() => bnum(_amount.value).gt(0));
+const hasBalance = computed(() => bnum(tokenBalance.value).gt(0));
+const isMaxed = computed(() => _amount.value === tokenBalance.value);
+
+const tokenBalance = computed(() => {
+  if (props.customBalance) return props.customBalance;
+  return balanceFor(_address.value);
+});
+
+const token = computed(() => {
+  if (!hasToken.value) return {};
+  return getToken(_address.value);
+});
+
+const tokenValue = computed(() => {
+  return toFiat(_amount.value, _address.value);
+});
+
+const rules = computed(() => {
+  if (!hasToken.value || !isWalletReady.value || props.noRules)
+    return [isPositive()];
+  return [
+    isPositive(),
+    isLessThanOrEqualTo(tokenBalance.value, t('exceedsBalance'))
+  ];
+});
+
+const maxPercentage = computed(() => {
+  if (!hasBalance.value || !hasAmount.value) return '0';
+
+  return bnum(_amount.value)
+    .div(tokenBalance.value)
+    .times(100)
+    .toFixed(1);
+});
+
+const barColor = computed(() => {
+  if (bnum(_amount.value).gt(tokenBalance.value)) return 'red';
+  return 'green';
+});
+
+const priceImpactSign = computed(() =>
+  (props.priceImpact || 0) >= 0 ? '-' : '+'
+);
```

Comment:
I don't think we actually know if the PI is positive. Did you not look into this @evgenyboxer? It seems that for negative price impact the SOR returns a positive number. So I assume this is an absolute number coming from the SOR. We may need to fix this.

For now I've assumed if the PI number coming out of the SOR is negative, it's actually a positive PI. I don't think the SOR will return a negative PI numbers so this should never happen. This will need to be refactored in a separate PR.

Reply:
I have looked into it and it *can* be a negative if we don't do the `Math.max` operation to get a "min" price impact.

However, sometimes that negative was actually a positive which got me confused... gotta follow this up with John. I think its great there is support for it in the UI.